tiledb/api/src/query_arrow/arrow.rs

```diff
@@ -0,0 +1,520 @@
+use std::sync::Arc;
+
+use arrow::datatypes as adt;
+
+use thiserror::Error;
+
+use crate::array::schema::CellValNum;
+use crate::datatype::Datatype;
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum Error {
+    #[error("Cell value size '{0}' is out of range.")]
+    CellValNumOutOfRange(u32),
+
+    #[error("Internal type error: Unhandled Arrow type: {0}")]
+    InternalTypeError(adt::DataType),
+
+    #[error("Invalid fixed sized length: {0}")]
+    InvalidFixedSize(i32),
+
+    #[error("Invalid Arrow type for conversion: '{0}'")]
+    InvalidTargetType(adt::DataType),
+
+    #[error("Failed to convert Arrow list element type: {0}")]
+    ListElementTypeConversionFailed(Box<Error>),
+
+    #[error(
+        "The TileDB datatype '{0}' does not have a default Arrow DataType."
+    )]
+    NoDefaultArrowType(Datatype),
+
+    #[error("Arrow type '{0} requires the TileDB field to be single valued.")]
+    RequiresSingleValued(adt::DataType),
+
+    #[error("Arrow type {0} requires the TileDB field be var sized.")]
+    RequiresVarSized(adt::DataType),
+
+    #[error("TileDB does not support timezones on timestamps")]
+    TimeZonesNotSupported,
+
+    #[error(
+        "TileDB type '{0}' and Arrow type '{1}' have different physical sizes"
+    )]
+    PhysicalSizeMismatch(Datatype, adt::DataType),
+
+    #[error("Unsupported Arrow DataType: {0}")]
+    UnsupportedArrowDataType(adt::DataType),
+
+    #[error("TileDB does not support lists with element type: '{0}'")]
+    UnsupportedListElementType(adt::DataType),
+
+    #[error("The Arrow DataType '{0}' is not supported.")]
+    ArrowTypeNotSupported(adt::DataType),
+    #[error("DataFusion does not support multi-value cells.")]
+    InvalidMultiCellValNum,
+    #[error("The TileDB Datatype '{0}' is not supported by DataFusion")]
+    UnsupportedTileDBDatatype(Datatype),
+    #[error("Variable-length datatypes as list type elements are not supported by TileDB")]
+    UnsupportedListVariableLengthElement,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// ConversionMode dictates whether certain conversions are allowed
+pub enum ConversionMode {
+    /// Only allow conversions that are semantically equivalent
+    Strict,
+    /// Allow conversions as long as the physical type is maintained.
+    Relaxed,
+}
+
+pub struct ToArrowConverter {
+    mode: ConversionMode,
+}
+
+impl ToArrowConverter {
+    pub fn strict() -> Self {
+        Self {
+            mode: ConversionMode::Strict,
+        }
+    }
+
+    pub fn physical() -> Self {
+        Self {
+            mode: ConversionMode::Relaxed,
+        }
+    }
+
+    pub fn convert_datatype(
+        &self,
+        dtype: &Datatype,
+        cvn: &CellValNum,
+        nullable: bool,
+    ) -> Result<adt::DataType> {
+        if let Some(arrow_type) = self.default_arrow_type(dtype) {
+            self.convert_datatype_to(dtype, cvn, nullable, arrow_type)
```

Comment:
This is great, the source/target type pairing is important but this also makes it not required. Very nice

Reply:
Though perhaps the `default_arrow_type` should also account for the cvn and nullability?

When requesting the default type, conversion to whatever the default is should always work.
When requesting a specific type, it's odd that `convert_datatype_to` can return something else. "Yes we can work with that" or "no way bro" seems like a more natural answer than "yes here's what you're actually going to get"

Reply:
I'm quite happy with how it separates the "pick what to convert to" from "check if conversion is legit" logic.

> it's odd that `convert_datatype_to` can return something else.

This one took me a second even after I went back and re-read that method. I totally glossed over the list/fixesizelist conversion before realizing that's what you must be referring to.

While writing this, I've started to think of the List/FixedList/Binary/Utf8 types as just an encoding of TileDB's CellValNum concept into the type system. As far as I'm concerned, these are basically non-negotiable parts of the conversion process. I.e., given a TileDB field with `Datatype::Int64, CellValNum(2)` its not really representable in anything other than a `FixedSizeList(fieldWithInt64type, 2)`. Same for `CellValNum::Var` going into a `List`. Now, we can obviously change the `Int64` to anything with a compatible width (i.e., any time type, UInt64, w/e), but the List/FixedSizeList modifiers can't change.

The only wrench in that logic is obviously the Binary/Utf8 dichotomy, since those are basically unrelated (type system wise) tough basically logically equivalent to `List`. My guess is that this is more engineering practicality than anything else. I.e., allowing for more efficient conversion to and from String types and the like rather than encoding that on top of the more abstract `List` interface.

Reply:
Selecting a bit from this discussion:
https://github.com/apache/datafusion/issues/11513

> some logically equivalent array types in the Arrow format have very different DataTypes – for example a logical string array can be represented in the Arrow array of DataType Utf8, Dictionary(Utf8), RunEndEncoded(Utf8), and StringView

Utf8 and List(u8) are physically the same but logically different because the first one requires valid UTF-8 and prevents construction if you don't have that.

> While writing this, I've started to think of the List/FixedList/Binary/Utf8 types as just an encoding of TileDB's CellValNum concept into the type system

Yep, this is how I see it too.  The initial arrow conversion layer I wrote ran with this, with the Exact/Inexact enums.  In retrospect LogicalMatch and PhysicalMatch might have been better names.

Reply:
> Utf8 and List(u8) are physically the same but logically different because the first one requires valid UTF-8 and prevents construction if you don't have that.

Huh, I must've said it in a different comment, but I totes agree there. I'm just vaguely surprised that `Utf8` doesn't wrap a `List(u8)` internally as they do for `OffsetsBuffer` wrapping a `ScalarBuffer::<i64>`. I.e., the `ScalarBuffer` is responsible for "Yep, these bytes work as a buffer of i64 values" and `OffsetsBuffer` then adds the "This buffer of i64 values is a valid list of non-negative monotonically increasing i64 values". I'm just assuming hand wavy String type compatibility or some such as the reason.

Reply:
I think the action I'm leaning towards here is actually to remove this function (convert) and replace with something like `is_physically_compatible` (not necessarily that but here I choose to be verbose).

Use case 1 - the user just asked for the field and we are allocating the buffer.  They will deal with the type later.  In this case there's no "conversion", we're just going to come up with some compatible type.

Use case 2 - the user supplies an arrow array and requests that we transfer data into it.  In this case we must check compatibility but there''s still no notion of a "conversion".

Use case 3 - the user requests an arrow data type but we allocate.  That's not implemented in this PR, it probably looks like item 2.